File: src/data/tutorials/033-maximum-depth-of-binary-tree.js
/**
 * Tutorial content for #033 — Maximum Depth of Binary Tree
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given the root of a binary tree, return its maximum depth — the number of nodes along the longest path from the root down to the farthest leaf node. A leaf is a node with no children.`,
    example: `Tree: root=3, left=9, right=20, 20.left=15, 20.right=7\n→ Path 3→9: length 2\n→ Path 3→20→15: length 3\n→ Path 3→20→7: length 3\n✅ Answer: 3`,
    keyInsight: `The maximum depth of a tree is 1 (for the current node) plus the maximum of the depths of its left and right subtrees. This recursive definition naturally leads to a DFS solution.`,
  },

  approaches: {
    'Recursive DFS': {
      intuition: `Ask the tree a simple recursive question: "What is my maximum depth?" The answer is: 0 if I'm null, otherwise 1 (for me) plus the bigger of my left and right subtree depths. This bottom-up recursion propagates the answer from leaves back up to the root.`,
      steps: [
        `Base case: if root is null, return 0.`,
        `Recursively compute leftDepth = maxDepth(root.left).`,
        `Recursively compute rightDepth = maxDepth(root.right).`,
        `Return 1 + max(leftDepth, rightDepth).`,
      ],
      example: `Tree: 3 → left:9, right:20 → 20.left:15, 20.right:7\n\nmaxDepth(9)  = 1 + max(0,0) = 1\nmaxDepth(15) = 1 + max(0,0) = 1\nmaxDepth(7)  = 1 + max(0,0) = 1\nmaxDepth(20) = 1 + max(1,1) = 2\nmaxDepth(3)  = 1 + max(1,2) = 3\n✅ Answer: 3`,
      keyInsight: `O(n) time (every node visited once), O(h) space for the call stack (h = tree height, O(n) worst case for a skewed tree). This is the most elegant and concise solution.`,
    },

    'BFS Level Order': {
      intuition: `Count levels using a BFS queue. Each time you finish processing all nodes at the current level, increment a depth counter. When the queue is empty you've processed all levels, so the counter equals the maximum depth.`,
      steps: [
        `If root is null, return 0.`,
        `Enqueue root. Set depth = 0.`,
        `While the queue is not empty: increment depth. Record levelSize = queue.length.`,
        `Process exactly levelSize nodes: dequeue each and enqueue their non-null children.`,
        `After the loop ends, return depth.`,
      ],
      example: `Tree: 3, left=9, right=20, 20.left=15, 20.right=7\n\nQueue: [3]        depth becomes 1, process 3, enqueue 9, 20\nQueue: [9, 20]    depth becomes 2, process 9 (no children), 20 (enqueue 15, 7)\nQueue: [15, 7]    depth becomes 3, process 15, 7 (no children)\nQueue: []\nReturn depth = 3\n✅ Answer: 3`,
      keyInsight: `O(n) time, O(n) space for the queue. Naturally counts depth without recursion — each BFS round corresponds to one tree level.`,
    },

    'Iterative DFS with Stack': {
      intuition: `Simulate the recursive DFS using an explicit stack. Push (node, depth) pairs. Track the maximum depth seen across all visited nodes. This avoids recursion entirely, which is useful for extremely deep trees.`,
      steps: [
        `If root is null, return 0.`,
        `Push (root, 1) onto a stack. Set maxDepth = 0.`,
        `While the stack is not empty: pop (node, depth).`,
        `Update maxDepth = max(maxDepth, depth).`,
        `If node.left exists, push (node.left, depth+1).`,
        `If node.right exists, push (node.right, depth+1).`,
        `Return maxDepth.`,
      ],
      example: `Tree: 3, left=9, right=20, 20.left=15, 20.right=7\n\nStack: [(3,1)]         maxDepth=0\nPop (3,1) maxDepth=1, push (9,2),(20,2)\nStack: [(9,2),(20,2)]  maxDepth=1\nPop (20,2) maxDepth=2, push (15,3),(7,3)\nPop (7,3)  maxDepth=3\nPop (15,3) maxDepth=3\nPop (9,2)  maxDepth=3\n✅ Answer: 3`,
      keyInsight: `O(n) time, O(h) space for the stack. Functionally identical to recursive DFS but iterative — avoids call stack overflow for very deep trees.`,
    },

    'Morris Traversal': {
      intuition: `Adapt the Morris threading technique to track depth without a stack or recursion. When descending left (creating a thread to a predecessor), increment a depth counter. When ascending back via a thread (removing it), subtract the number of steps used to find the predecessor — this restores the correct current depth. Track the maximum depth seen at each node visit.`,
      steps: [
        `Set curr = root, maxDepth = 0, currDepth = 0.`,
        `While curr is not null:`,
        `  If curr has no left child: increment currDepth, update maxDepth, move curr = curr.right.`,
        `  Otherwise, find the inorder predecessor (rightmost in left subtree), counting steps taken.`,
        `  If predecessor.right is null: create thread (predecessor.right = curr), increment currDepth, move curr = curr.left.`,
        `  If predecessor.right is curr: remove thread (predecessor.right = null), subtract steps from currDepth, move curr = curr.right.`,
      ],
      example: `Tree: 3 (root), left=9, right=20, 20.left=15, 20.right=7\n\ncurr=3: has left(9). Pred of 3 = node 9 (steps=1). Thread: 9.right→3. currDepth=1, go left→9.\ncurr=9: no left. currDepth=2, maxDepth=2. Move right (via thread)→3.\ncurr=3: has left(9). Pred=9, thread exists. Remove thread. currDepth-=1=1. Move right→20.\ncurr=20: has left(15). Pred=15 (steps=1). Thread: 15.right→20. currDepth=2, go left→15.\ncurr=15: no left. currDepth=3, maxDepth=3. Move right (via thread)→20.\ncurr=20: pred=15, thread exists. Remove. currDepth-=1=2. Move right→7.\ncurr=7: no left. currDepth=3, maxDepth=3. Move right→null. Done.\n✅ Answer: 3`,
      keyInsight: `O(n) time, O(1) space. Applying Morris traversal to depth calculation requires careful step-counting when threads are created and removed — currDepth must decrease by exactly the steps traversed to find the predecessor when the thread is unwound.`,
    },

    'Tail Recursion with Accumulator': {
      intuition: `Instead of returning depth bottom-up, pass the current depth downward as an accumulator parameter. At every node visit, compare the accumulated depth against a running maximum (stored as a field or result holder). Recurse on both children with depth+1. When the recursion finishes, the field holds the answer. This is effectively a pre-order DFS — visit the node first, then go deeper.`,
      steps: [
        `Initialise a class-level (or closure) variable maxAcc = 0.`,
        `Define helper(node, depth): if node is null, return immediately.`,
        `Update maxAcc = max(maxAcc, depth).`,
        `Recursively call helper(node.left, depth + 1) and helper(node.right, depth + 1).`,
        `Call helper(root, 1) and return maxAcc.`,
      ],
      example: `Tree: 3, left=9, right=20, 20.left=15, 20.right=7\n\nhelper(3, 1): maxAcc=1\n  helper(9, 2): maxAcc=2\n    helper(null, 3): return\n    helper(null, 3): return\n  helper(20, 2): maxAcc=2\n    helper(15, 3): maxAcc=3\n      helper(null,4): return × 2\n    helper(7, 3): maxAcc=3\n      helper(null,4): return × 2\nReturn maxAcc = 3\n✅ Answer: 3`,
      keyInsight: `O(n) time, O(h) space for the call stack. Functionally equivalent to recursive DFS but the depth flows down rather than bubbling up — avoids the final "1 + max(left, right)" computation at each node. Useful when the accumulator pattern is clearer to the reader.`,
    },
  },
}

File: public/cpp/033-maximum-depth-of-binary-tree.cpp
/*
 * LeetCode Problem #104: Maximum Depth of Binary Tree
 * Difficulty: Easy
 * Link: https://leetcode.com/problems/maximum-depth-of-binary-tree/
 * 
 * Problem Statement:
 * Given the root of a binary tree, return its maximum depth.
 * 
 * A binary tree's maximum depth is the number of nodes along the longest path
 * from the root node down to the farthest leaf node.
 * 
 * Example:
 * Input: root = [3,9,20,null,null,15,7]
 *     3
 *    / \
 *   9  20
 *     /  \
 *    15   7
 * Output: 3
 */

#include <queue>
#include <algorithm>
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    // ==================== APPROACH 1: Recursive DFS ====================
    /*
     * Algorithm:
     * - Recursively calculate depth of left and right subtrees
     * - Return 1 + max(left_depth, right_depth)
     * - Base case: null node has depth 0
     * 
     * Time Complexity: O(n) - visit each node once
     * Space Complexity: O(h) - recursion stack, h = height
     * 
     * When to use: This is the STANDARD solution
     * 
     * Key Insight:
     * - Depth of tree = 1 + max depth of subtrees
     * - Natural recursive definition
     * - Clean and elegant
     */
    int maxDepth(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        
        int leftDepth = maxDepth(root->left);
        int rightDepth = maxDepth(root->right);
        
        return 1 + max(leftDepth, rightDepth);
    }
    
    // ==================== APPROACH 2: BFS (Level Order) ====================
    /*
     * Algorithm:
     * - Use queue for level-order traversal
     * - Count number of levels
     * - Each level increments depth
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(w) - w is max width
     * 
     * When to use: When iterative solution preferred
     * 
     * Key Insight:
     * - BFS naturally processes level by level
     * - Number of levels = depth
     * - No recursion needed
     */
    int maxDepth_BFS(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        
        queue<TreeNode*> q;
        q.push(root);
        int depth = 0;
        
        while (!q.empty()) {
            int levelSize = q.size();
            depth++;
            
            for (int i = 0; i < levelSize; i++) {
                TreeNode* node = q.front();
                q.pop();
                
                if (node->left != nullptr) {
                    q.push(node->left);
                }
                if (node->right != nullptr) {
                    q.push(node->right);
                }
            }
        }
        
        return depth;
    }
    
    // ==================== APPROACH 3: DFS with Stack (Iterative) ====================
    /*
     * Algorithm:
     * - Use stack to simulate recursion
     * - Track depth for each node
     * - Update maximum depth seen
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(h)
     * 
     * When to use: When recursion not allowed
     * 
     * Key Insight:
     * - Explicit stack replaces recursion
     * - Track depth alongside nodes
     * - Preorder traversal approach
     */
    int maxDepth_DFS_Stack(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        
        stack<pair<TreeNode*, int>> stk;
        stk.push({root, 1});
        int maxDepth = 0;
        
        while (!stk.empty()) {
            auto [node, depth] = stk.top();
            stk.pop();
            
            maxDepth = max(maxDepth, depth);
            
            if (node->right != nullptr) {
                stk.push({node->right, depth + 1});
            }
            if (node->left != nullptr) {
                stk.push({node->left, depth + 1});
            }
        }
        
        return maxDepth;
    }
    
    // ==================== APPROACH 4: Morris Traversal ====================
    /*
     * Algorithm:
     * - Use Morris traversal for O(1) space
     * - Track depth during traversal
     * - No recursion or stack needed
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     * 
     * When to use: When O(1) space required
     * 
     * Key Insight:
     * - Threaded binary tree concept
     * - Temporarily modify tree structure
     * - Restore tree after traversal
     */
    int maxDepth_Morris(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        
        int maxDepth = 0;
        int currentDepth = 0;
        TreeNode* current = root;
        
        while (current != nullptr) {
            if (current->left == nullptr) {
                currentDepth++;
                maxDepth = max(maxDepth, currentDepth);
                current = current->right;
            } else {
                // Find predecessor
                TreeNode* predecessor = current->left;
                int steps = 1;
                
                while (predecessor->right != nullptr && 
                       predecessor->right != current) {
                    predecessor = predecessor->right;
                    steps++;
                }
                
                if (predecessor->right == nullptr) {
                    // Create thread
                    currentDepth++;
                    predecessor->right = current;
                    current = current->left;
                } else {
                    // Remove thread
                    predecessor->right = nullptr;
                    currentDepth -= steps;
                    current = current->right;
                }
            }
        }
        
        return maxDepth;
    }
    
    // ==================== APPROACH 5: Tail Recursion Optimized ====================
    /*
     * Algorithm:
     * - Use tail recursion with accumulator
     * - Pass current depth as parameter
     * - Update maximum in each call
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(h)
     * 
     * When to use: For tail recursion optimization
     * 
     * Key Insight:
     * - Tail recursion can be optimized by compiler
     * - Accumulator pattern
     * - Still uses stack space in practice
     */
    int maxDepth_TailRecursion(TreeNode* root) {
        int result = 0;
        maxDepthHelper(root, 1, result);
        return result;
    }
    
private:
    void maxDepthHelper(TreeNode* node, int depth, int& maxDepth) {
        if (node == nullptr) {
            return;
        }
        
        maxDepth = max(maxDepth, depth);
        maxDepthHelper(node->left, depth + 1, maxDepth);
        maxDepthHelper(node->right, depth + 1, maxDepth);
    }
};

// ==================== TEST CASES ====================
void runTests() {
    Solution sol;
    
    // Test Case 1: Standard tree
    // Input: [3,9,20,null,null,15,7]
    //     3
    //    / \
    //   9  20
    //     /  \
    //    15   7
    // Expected: 3
    cout << "Test Case 1: Standard tree\n";
    TreeNode* root1 = new TreeNode(3);
    root1->left = new TreeNode(9);
    root1->right = new TreeNode(20);
    root1->right->left = new TreeNode(15);
    root1->right->right = new TreeNode(7);
    cout << "Output: " << sol.maxDepth(root1) << "\n";
    cout << "Expected: 3\n\n";
    
    // Test Case 2: Single node
    // Input: [1]
    // Expected: 1
    cout << "Test Case 2: Single node\n";
    TreeNode* root2 = new TreeNode(1);
    cout << "Output: " << sol.maxDepth_BFS(root2) << "\n";
    cout << "Expected: 1\n\n";
    
    // Test Case 3: Empty tree
    // Input: []
    // Expected: 0
    cout << "Test Case 3: Empty tree\n";
    TreeNode* root3 = nullptr;
    cout << "Output: " << sol.maxDepth_DFS_Stack(root3) << "\n";
    cout << "Expected: 0\n\n";
    
    // Test Case 4: Left-skewed tree
    // Input: [1,2,null,3]
    //   1
    //  /
    // 2
    // /
    //3
    // Expected: 3
    cout << "Test Case 4: Left-skewed tree\n";
    TreeNode* root4 = new TreeNode(1);
    root4->left = new TreeNode(2);
    root4->left->left = new TreeNode(3);
    cout << "Output: " << sol.maxDepth_Morris(root4) << "\n";
    cout << "Expected: 3\n\n";
    
    // Test Case 5: Right-skewed tree
    // Input: [1,null,2,null,3]
    // 1
    //  \
    //   2
    //    \
    //     3
    // Expected: 3
    cout << "Test Case 5: Right-skewed tree\n";
    TreeNode* root5 = new TreeNode(1);
    root5->right = new TreeNode(2);
    root5->right->right = new TreeNode(3);
    cout << "Output: " << sol.maxDepth_TailRecursion(root5) << "\n";
    cout << "Expected: 3\n\n";
}

int main() {
    runTests();
    return 0;
}

/*
 * COMPARISON OF APPROACHES:
 * 
 * 1. Recursive DFS (RECOMMENDED):
 *    Time: O(n), Space: O(h)
 *    Pros: Clean, elegant, natural
 *    Cons: Stack overflow for very deep trees
 *    Best for: Most interview scenarios
 * 
 * 2. BFS Level Order:
 *    Time: O(n), Space: O(w)
 *    Pros: Iterative, counts levels directly
 *    Cons: More space for wide trees
 *    Best for: When iterative preferred
 * 
 * 3. DFS with Stack:
 *    Time: O(n), Space: O(h)
 *    Pros: No recursion
 *    Cons: More complex than recursive
 *    Best for: When recursion not allowed
 * 
 * 4. Morris Traversal:
 *    Time: O(n), Space: O(1)
 *    Pros: Constant space
 *    Cons: Complex, temporarily modifies tree
 *    Best for: When O(1) space required
 * 
 * 5. Tail Recursion:
 *    Time: O(n), Space: O(h)
 *    Pros: Tail recursion pattern
 *    Cons: Not truly optimized in most languages
 *    Best for: Academic interest
 * 
 * INTERVIEW TIPS:
 * - Start with recursive solution (simplest)
 * - Explain base case clearly
 * - Discuss recursion tree
 * - Mention BFS alternative
 * - Consider space complexity
 * - Handle edge cases: null, single node
 * - Discuss balanced vs skewed trees
 * - Mention Morris for O(1) space
 * - Consider follow-up: minimum depth
 * - Discuss height vs depth terminology
 * 
 * KEY INSIGHTS:
 * - Depth = 1 + max(left_depth, right_depth)
 * - Recursive definition is natural
 * - BFS counts levels directly
 * - Space depends on tree shape
 * - Balanced tree: O(log n) space
 * - Skewed tree: O(n) space
 * 
 * STEP-BY-STEP WALKTHROUGH:
 * Tree: [3,9,20,null,null,15,7]
 *     3
 *    / \
 *   9  20
 *     /  \
 *    15   7
 * 
 * maxDepth(3):
 *   leftDepth = maxDepth(9):
 *     leftDepth = maxDepth(null) = 0
 *     rightDepth = maxDepth(null) = 0
 *     return 1 + max(0,0) = 1
 *   
 *   rightDepth = maxDepth(20):
 *     leftDepth = maxDepth(15):
 *       return 1 + max(0,0) = 1
 *     rightDepth = maxDepth(7):
 *       return 1 + max(0,0) = 1
 *     return 1 + max(1,1) = 2
 *   
 *   return 1 + max(1,2) = 3
 * 
 * COMMON MISTAKES:
 * - Confusing depth with height
 * - Not handling null root
 * - Off-by-one errors in counting
 * - Forgetting base case
 * - Not considering skewed trees
 * - Stack overflow with deep recursion
 * - Incorrect BFS level counting
 * - Memory leaks from not deleting nodes
 * - Confusing max with min depth
 * - Not considering empty tree
 * 
 * FOLLOW-UP QUESTIONS:
 * - How would you find minimum depth?
 * - How would you check if tree is balanced?
 * - Can you do it with O(1) space?
 * - How would you find diameter of tree?
 * - What if tree is n-ary instead of binary?
 * 
 * RELATED PROBLEMS:
 * - LeetCode #111: Minimum Depth of Binary Tree
 * - LeetCode #110: Balanced Binary Tree
 * - LeetCode #543: Diameter of Binary Tree
 * - LeetCode #559: Maximum Depth of N-ary Tree
 * - LeetCode #662: Maximum Width of Binary Tree
 */

// Made with Bob
